pkg/resource/target_reconciler.go

```diff
@@ -0,0 +1,124 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+)
+
+const (
+	targetControllerName   = "kubernetestarget.crossplane.io"
+	targetReconcileTimeout = 1 * time.Minute
+
+	errGetTarget                 = "unable to get Target"
+	errMissingSecretRef          = "no secretRef specified for Target with a clusterRef"
+	errManagedResourceIsNotBound = "managed resource in Target clusterRef is unbound"
+	errUpdateTarget              = "unable to update Target"
+)
+
+// A TargetKind contains the type metadata for a kind of target resource.
+type TargetKind schema.GroupVersionKind
+
+// A TargetReconciler reconciles targets by propagating the secret of the
+// referenced managed resource.
+type TargetReconciler struct {
+	client     client.Client
+	newTarget  func() Target
+	newManaged func() Managed
+
+	propagator ManagedConnectionPropagator
+}
+
+// NewTargetReconciler returns a Reconciler that reconciles
+// KubernetesTarget by setting their ConnectionSecretRef to that of the
+// referenced Kubernetes cluster implementation.
+func NewTargetReconciler(m manager.Manager, of TargetKind, with ManagedKind) *TargetReconciler {
+	nt := func() Target { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Target) }
+	nr := func() Managed { return MustCreateObject(schema.GroupVersionKind(with), m.GetScheme()).(Managed) }
+
+	// Panic early if we've been asked to reconcile a target or resource kind
+	// that has not been registered with our controller manager's scheme.
+	_, _ = nt(), nr()
+
+	r := &TargetReconciler{
+		client:     m.GetClient(),
+		newTarget:  nt,
+		newManaged: nr,
+		propagator: NewAPIManagedConnectionPropagator(m.GetClient(), m.GetScheme()),
+	}
+
+	return r
+}
+
+// Reconcile a target with a concrete managed resource.
+func (r *TargetReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("Reconciling", "controller", targetControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), targetReconcileTimeout)
+	defer cancel()
+
+	target := r.newTarget()
+	if err := r.client.Get(ctx, req.NamespacedName, target); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetTarget)
+	}
+
+	if target.GetWriteConnectionSecretToReference() == nil {
```

Comment:
Not sure if we should use predicates to filter out any targets that do not have their `SecretRef` field set? I kind of like how this tells the user why the secret is not being propagated.

Reply:
It seems like we're going with the approach in which we don't automatically populate this reference when a cluster is being referenced? i.e. The user must specify both the target and where to write the secret? If so I think this makes sense.

Reply:
Yes, that is where we stand currently. The tricky part here is that `writeConnectionSecretToRef` would make sense in the case that a cluster is referenced, but just `connectionSecretRef` makes sense if the secret is being referenced directly (i.e. you have manually created a secret and are referencing it). We could also have both of these fields and then...
1. If `connectionSecretRef` is set then we ignore `writeConnectionSecretToRef` and `clusterRef` and always assume that you have manually created the secret.
1. If `writeConnectionSecretToRef` is set and `clusterRef` is not then we do reconcile with the shared Target reconciler because the missing `clusterRef` means that the Target will not pass predicates of any controllers.
1. If `clusterRef` is set but `writeConnectionSecretToRef` is not then we hit this situation in the code above where we say "please tell us where to write your secret to" in the Target status. We could also opt to generate the secret name here?

What do you think about this pattern?

Reply:
Having two different ways to reference a secret seems more complex than it needs to be to me. My preference is to just name this field `connectionSecretRef` and use it for both cases.

Reply:
It is a local secret so we would only need name, but I generally agree that we should generate the name if not provided (per 3 above).

Reply:
Was looking at this and am wondering if we have any other examples of updating a `spec` field of a user-created managed resource? Feels like a bit of an anti-pattern.

Reply:
Any of our managed resource late initializations are an example of this; upstream types like `Service` do it to (e.g. will set an IP if you don't specify one).

Reply:
Good point. Are you in agreement on generating secret name of not provided?

Reply:
I think:

* If a secret is specified (either by the user or by us having propagated one from a specified target) we ignore the target (if any) and use the specified secret.
* If a target is specified _and_ no secret is specified, we create a secret and configure the target's secret to propagate to our newly created secret.

This means that in the case where a user specifies a target _and_ a secret, we ignore the target and just use the secret.

Reply:
I am assuming you mean Cluster where you use Target here, and if so I mostly agree, but it would mean that the case where both are provided we would not propagate the secret. I think if both the cluster is referenced and a secret name is provided then we should propagate to that secret name.

Reply:
> I am assuming you mean Cluster where you use Target here

Derp - yeah I mean cluster.

> I think if both the cluster is referenced and a secret name is provided then we should propagate to that secret name.

So: 

1. If only a secret is specified, we presume it exists and use it to connect to the cluster.
2. If only a cluster is specified, we generate a secret name and setup propagation.
3. If both are specified, we setup propagation from the cluster to the secret.

For some reason I was thinking 3 would not work, but now in the future of 10 minutes later I can't recall why. I imagine it won't work if the user specifies an _existing_ secret (which won't have the propagation annotations), but if they specify a non-existent secret we'll be able to create it.

Reply:
I *think* you are right about the case in which the `Secret` already exists because we will try to get a `Secret` of the supplied name but it will not be owned by the `Target` so I imagine `CreateOrUpdate` will fail. I think this is the desired behavior we want here.